_posts/2023-8-22-introducing-sqlite3-to-kw.md

```diff
@@ -0,0 +1,114 @@
+
+The structure described above is my understanding of a file-based database. At least, this was the structure present in kw.
+
+## kw old database
+
+> The following description is based on the `unstable` branch at commit `#a42592a`. You can check kw's repo at this state [here](https://github.com/kworkflow/kworkflow/tree/a42592a5fa7c6704d62f3d08f2486d1964223887).
+
+As an [XDG-compliant application](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), kw stores its **user-specific data files** at `~/local/share/kw`. In this sense, there were three sub-directories `~/local/share/kw/statistics`, `~/local/share/kw/pomodoro`, and `~/local/share/kw/configs` that functioned like databases. The first stored files related to any statistic collected by kw. The second stored files related to Pomodoro sessions (from [`kw pomodoro`](https://kworkflow.org/man/features/pomodoro.html)). The third one stored Linux kernel `.config` files and metadata for the [`kw kernel-config-manager`](https://kworkflow.org/man/features/kernel-config-manager.html) feature.
+
+For statistics, a file `statistics/<year>/<month>/<day>` represented statistics collected at _\<month\>/\<day\>/\<year\>_. For example, a line
+
+```
+build 497
+```
+
+in a file `statistics/23/08/23`, meant that a `kw build` command ran on August 23 of 2023 and lasted for 8 minutes and 17 seconds (497 seconds).
+
+`kw pomodoro` had this same file structure that represented dates, with each line representing an entry. Differently from the statistics database though, each line/entry was comma separated, had a different number of attributes, and also had an optional attribute. On top of that, there was a file `~/local/share/kw/pomodoro/tags` for storing Pomodoro tags and a file `~/local/share/kw/pomodoro_current.log` for the active Pomodoro timeboxes.
+
+I could also explain the intricacies of the `kw kernel-config-manager` database (which had even more particularities), but it would probably be tiresome for you the reader.
+
+The point may be already clear: although functional, each feature had to implement its own database with its own details. This made the code hard to scale and more coupled with these particularities of **where** and **how** the data was stored.
+
+## The right DBMS
+
+DBMSs are really vast and diverse, proposing different solutions for different problems. The people involved in kw knew that the introduction of a DBMS was necessary and agreed on some requisites for the system:
+
+- Be _Free Libre and Open Source Software_ (FLOSS).
+- Have a CLI interface for easy integration with Bash script, as we want to maintain kw's codebase in pure bash wherever possible.
+- Have a small footprint.
+- Run on user space.
+- Be a Relational DBMS.
```

Comment:
It also needs to be portable, something easy to set up. We could use Postgresql, but it has more setup overhead than SQLite.